### PR TITLE
Feature/sre 89

### PR DESCRIPTION
**Pull request by E-ricus**
## JIRA Ticket URL
https://merlinapp.atlassian.net/browse/SRE-89

## Description
Create kafka Publisher with Avro schema support

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I did add reviewers to this PR
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I created the unit tests and they are passing

### CODE REVIEW

**Review by pcu4dros — APPROVED** [review]

File: kafka/avro.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+		value = val
+	}
+	f := Field{
+		Name: key,
+		Type: value,
+	}
+	return f
+}
+
+func createComplexField(val interface{}) interface{} {
```
Comment (daltemen) [inline kafka/avro.go]: a docstring would improve the readability of the method

File: kafka/avro.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+			if val, ok := v[1].([]interface{}); ok {
+				return map[string]interface{}{"type": t, "items": createComplexField(val)}
+			}
+			return map[string]interface{}{"type": t, "items": createSchema(t, v[1].(map[string]interface{}))}
+		}
+	case "map":
+		if ok {
+			return map[string]string{"type": t, "values": tt}
+		} else {
+			return map[string]interface{}{"type": t, "values": createComplexField(v[1])}
```
Comment (daltemen) [inline kafka/avro.go]: I would think that creating data types can improve the method a little

File: kafka/publisher.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+func NewPublisher(kafkaServerAddress, schemaServerAddress, topic string, objectType interface{}) queuesgo.Publisher {
+	if !queuesgo.ValidateType(objectType) {
+		return nil
+	}
+	schema := createSchema(queuesgo.GetName(objectType), queuesgo.GetFields(objectType))
+	schemaBytes, err := json.Marshal(schema)
+	if err != nil {
+		return nil
+	}
+	fmt.Println("Schema registered: " + string(schemaBytes))
```
Comment (daltemen) [inline kafka/publisher.go:38]: fmt print necessary ?

**Review by daltemen — APPROVED** [review]
minors

File: kafka/publisher.go (hunk 3 above)
Comment (E-ricus) [reply kafka/publisher.go:38]: Yes! this is the schema registered on the schema registry server if the topic didn't exist
if that's not the case, then the schema should be exactly the same as the registered this helps a lot to verify

File: kafka/avro.go (hunk 1 above)
Comment (E-ricus) [reply kafka/avro.go]: Agree, the explication of the complex fields is on the docstring of the GetFields function, but a specific one would help as the reflection complexity is high
I'll add it on the next PR :)

File: kafka/avro.go (hunk 2 above)
Comment (E-ricus) [reply kafka/avro.go]: Agree, I'll add it with the docstring on the next PR

Comment (E-ricus) [reply kafka/avro.go]: I finally added it here

File: kafka/avro.go (hunk 1 above)
Comment (E-ricus) [reply kafka/avro.go]: I finally added it here

File: kafka/avro.go [hunk 4]
```diff
@@ -15,6 +15,16 @@ type Field struct {
...
+type ListField struct {
+	Type  string      `json:"type"`
+	Items interface{} `json:"items"`
+}
+
+type MapField struct {
+	Type   string      `json:"type"`
+	Values interface{} `json:"values"`
+}
+
```
Comment (daltemen) [inline kafka/avro.go:27]: 👍

**Review by daltemen — APPROVED** [review]

**Review by pcu4dros — APPROVED** [review]

**Review by dfsanchezg — APPROVED** [review]

**Review by oscanezing — APPROVED** [review]

**Pull request merged**